Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp

```diff
@@ -0,0 +1,444 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "MemoryMappedGPUBuffer.h"
+
+#if USE(GBM)
+#include "DRMDeviceManager.h"
+#include "DRMDeviceNode.h"
+#include "GBMVersioning.h"
+#include "IntRect.h"
+#include "Logging.h"
+#include "PlatformDisplay.h"
+
+#include <epoxy/egl.h>
+#include <linux/dma-buf.h>
+#include <sys/ioctl.h>
+#include <wtf/SafeStrerror.h>
+#include <wtf/StdLibExtras.h>
+#include <wtf/unix/UnixFileDescriptor.h>
+
+#if USE(LIBDRM)
+#include <drm_fourcc.h>
+#include <xf86drm.h>
+#endif
+
+namespace WebCore {
+
+constexpr bool kDumpSupportedDRMFormatsAndModifiers = false;
+constexpr bool kDumpBufferFormatNegotiation = false;
+constexpr bool kDumpBufferAllocation = false;
+constexpr bool kUseDRMModifiers = true;
+
+MemoryMappedGPUBuffer::MemoryMappedGPUBuffer(const IntSize& size)
+    : m_size(size)
+{
+}
+
+MemoryMappedGPUBuffer::~MemoryMappedGPUBuffer()
+{
+    unmapIfNeeded();
+
+    if (m_bo)
+        gbm_bo_destroy(m_bo);
+}
+
+std::unique_ptr<MemoryMappedGPUBuffer> MemoryMappedGPUBuffer::create(const IntSize& size, OptionSet<BufferFlag> flags)
+{
+    auto& manager = WebCore::DRMDeviceManager::singleton();
+    ASSERT(manager.isInitialized());
+
+    auto deviceNode = manager.mainDeviceNode(WebCore::DRMDeviceManager::NodeType::Render);
+    if (!deviceNode) {
+        LOG(GPUBuffers, "MemoryMappedGPUBuffer::create(), failed to get GBM render device node");
+        return nullptr;
+    }
+
+    LOG(GPUBuffers, "MemoryMappedGPUBuffer::create(), allocating dma-bufs from device '%s'", deviceNode->filename().data());
+
+    auto* device = deviceNode->gbmDevice();
+    if (!device) {
+        LOG(GPUBuffers, "MemoryMappedGPUBuffer::create(), failed to get GBM device");
+        return nullptr;
+    }
+
+    constexpr uint32_t preferredDMABufFormat = DRM_FORMAT_ABGR8888;
+
+    const auto& supportedFormats = PlatformDisplay::sharedDisplay().dmabufFormats();
+    if constexpr (kDumpSupportedDRMFormatsAndModifiers) {
```

Comment:
I was looking into the loops and thinking about avoiding the impact if not logging, then I checked properly and your approach is very good -- a constexpr check. I like this a lot!